### PR TITLE
[LO-1769] Fix default option for anyOf fields

### PR DESCRIPTION
**Pull request by jdmunro**
There's a bug in the form library that can be reproduced as following in teach:

1. Create an exercise with a flavor - change the flavor and partially edit the data
2. Swap to any other exercise
3. Navigate back to the first exercise
4. The flavor has now reverted to the default one, and the form is incorrect

This happens because when we navigate away from the page, the internal state of the form is lost. It falls back to the props we pass in, but it attempts to validate these with the schema to find the right selected option. However, since we have only partially filled the form, it's going to fail.

The solution is to recursively remove the `required` field from a copy of the schema. They already had a hack for this on the top-level, but we need to do it recursively also.

### CODE REVIEW

File: packages/core/test/utils_test.js [hunk 1]
```diff
@@ -3714,4 +3715,173 @@ describe("utils", () => {
...
+      it("returns the expected option if it matches the schema", () => {
+        const data = { type: "Two" };
+
+        expect(getMatchingOption(data, options)).eql(1);
+      });
+    });
+
+    describe("partial form data", () => {
+      it("returns the expected option for a partial match", () => {
+        const options = Object.freeze([
```
Comment (jdmunro) [inline packages/core/test/utils_test.js:3761]: I used `Object.freeze` because we need to make sure we don't mutate the original schema by accident

Comment (jdmunro) [reply packages/core/test/utils_test.js:3761]: (it throws an exception if you mutate the object)

Screenshot 2020-12-02 at 15 15 18

**Review by rv2e — APPROVED** [review]
🙈

**Comment by rv2e** [comment]
Great that you've added tests by the way :)

**Comment by entropitor** [comment]
Should we try to get this in the public PR as well?

**Comment by jdmunro** [comment]
@entropitor Yeah I will look at this - though @hermansje hasn't had any luck with his PR (the repo has 108 open PRs 😅) so I wouldn't get our hopes up. As I understand it, they are working on a v3 which might change a few things.

**Pull request merged**